common/include/camera/ICameraControl.hpp

```diff
@@ -0,0 +1,312 @@
+        static_assert(NblVirtualKeys.size() == EventsCount);
+
+        for (const auto virtualKey : NblVirtualKeys)
+        {
+            const auto code = m_keysToEvent[virtualKey];
+            bool& keyDown = m_keysDown[virtualKey];
+
+            using virtual_key_state_t = std::tuple<ui::E_KEY_CODE /*physical key representing virtual key*/, bool /*is pressed*/, float64_t /*delta action*/>;
+
+            auto updateVirtualState = [&]() -> virtual_key_state_t
+            {
+                virtual_key_state_t state = { ui::E_KEY_CODE::EKC_NONE, false, 0.f };
+
+                for (const auto& ev : events) // TODO: improve the search
+                {
+                    if (ev.keyCode == code)
+                    {
+                        if (ev.action == nbl::ui::SKeyboardEvent::ECA_PRESSED && !keyDown)
+                            keyDown = true;
+                        else if (ev.action == nbl::ui::SKeyboardEvent::ECA_RELEASED)
+                            keyDown = false;
+
+                        const auto dt = std::chrono::duration_cast<std::chrono::milliseconds>(m_nextPresentationTimeStamp - ev.timeStamp).count();
+                        assert(dt >= 0);
+
+                        state = std::make_tuple(code, keyDown, dt);
+                        break;
+                    }
+                }
+
+                return state;
+            };
+
+            const auto&& [physicalKey, isDown, dtAction] = updateVirtualState();
+
+            if (physicalKey != ui::E_KEY_CODE::EKC_NONE)
+                if (isDown)
+                    output.emplace_back(CVirtualEvent{ virtualKey, static_cast<float64_t>(dtAction) });
+        }
+
+        return output;
+    }
+
+    /*
+    // [OPTIONAL]: process mouse to generate virtual manipulation events
+    // note that:
+    // - all manipulations *may* be done with keyboard keys (if you have a touchpad or sth an ui:: event could be a code!)
+    // - it doesn't make the manipulation itself!
+    */
+    std::vector<CVirtualEvent> processMouse(std::span<const ui::SMouseEvent> events) const
+    {
+        double dPitch = {}, dYaw = {};
+
+        for (const auto& ev : events)
+            if (ev.type == nbl::ui::SMouseEvent::EET_MOVEMENT)
+            {
+                dYaw += ev.movementEvent.relativeMovementX;
+                dPitch += ev.movementEvent.relativeMovementY;
+            }
+
+        std::vector<CVirtualEvent> output;
+
+        if (dPitch)
+        {
+            auto& pitch = output.emplace_back();
+            pitch.type = dPitch > 0.f ? TiltUp : TiltDown;
+            pitch.value = std::abs(dPitch);
+        }
+
+        if (dYaw)
+        {
+            auto& yaw = output.emplace_back();
+            yaw.type = dYaw > 0.f ? PanRight : PanLeft;
+            yaw.value = std::abs(dYaw);
+        }
```

Comment:
this seems very "hardwired" to how FPS camera works

Reply:
the semantic indeed suggest FPS camera like stuff, however the impulses we return can be used to implement any type of camera, need to think about this but mouse events at the end are just XY delta motion